src/records/RecYAMLDecoder.cc

```diff
@@ -0,0 +1,213 @@
+/** @file
+
+  Decode the records.yaml configuration file.
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#include "records/P_RecCore.h"
+#include "records/RecYAMLDecoder.h"
+#include "records/RecYAMLDefs.h"
+
+#include "tscore/Diags.h"
+
+#include "RecordsConfig.h"
+
+#include <string_view>
+
+#include <swoc/Errata.h>
+#include <swoc/BufferWriter.h>
+
+static std::array<std::string_view, 5> Node_Type_to_Str{{"Undefined", "Null", "Scalar", "Sequence", "Map"}};
+namespace
+{
+constexpr std::string_view CONFIG_RECORD_PREFIX{"proxy.config"};
+const inline std::string RECORD_YAML_ROOT_STR{"ts"};
+
+// I will move this to a common place.
+constexpr std::string_view YAML_FLOAT_TAG_URI{"tag:yaml.org,2002:float"};
+constexpr std::string_view YAML_INT_TAG_URI{"tag:yaml.org,2002:int"};
+constexpr std::string_view YAML_STR_TAG_URI{"tag:yaml.org,2002:str"};
+constexpr std::string_view YAML_BOOL_TAG_URI{"tag:yaml.org,2002:bool"};
+constexpr std::string_view YAML_NULL_TAG_URI{"tag:yaml.org,2002:null"};
+
+} // namespace
+
+namespace detail
+{
+void flatten_node(CfgNode const &field, RecYAMLNodeHandler handler, swoc::Errata &errata);
+std::pair<RecDataT, std::string> try_deduce_type(YAML::Node const &node);
+
+// Helper class to make the code less verbose when lock is needed.
+struct scoped_cond_lock {
```

Comment:
Gah. Better to switch to `std::mutex` with `std::shared_lock` and `std::unique_lock`.